packages/lit-html/src/directives/async-append.ts

```diff
@@ -4,97 +4,43 @@
  */
 
-import {ChildPart, noChange} from '../lit-html.js';
+import {ChildPart} from '../lit-html.js';
 import {
   directive,
   DirectiveParameters,
   PartInfo,
   PartType,
 } from '../directive.js';
-import {AsyncDirective} from '../async-directive.js';
+import {AsyncReplaceDirective} from './async-replace.js';
 import {
   clearPart,
   insertPart,
   setChildPartValue,
 } from '../directive-helpers.js';
 
-type Mapper<T> = (v: T, index?: number) => unknown;
+class AsyncAppendDirective extends AsyncReplaceDirective {
+  private __childPart!: ChildPart;
 
-class AsyncAppendDirective extends AsyncDirective {
-  private _value?: AsyncIterable<unknown>;
-  private _reconnectResolver?: () => void;
-  private _reconnectPromise?: Promise<void>;
-
-  constructor(partInfo: PartInfo) {
-    super(partInfo);
+  protected validatePartType(partInfo: PartInfo) {
     if (partInfo.type !== PartType.CHILD) {
       throw new Error('asyncAppend can only be used in child expressions');
     }
   }
 
-  // @ts-expect-error value not used, but we want a nice parameter for docs
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  render<T>(value: AsyncIterable<T>, _mapper?: Mapper<T>) {
-    return noChange;
+  update(part: ChildPart, params: DirectiveParameters<this>) {
+    this.__childPart = part;
+    return super.update(part, params);
   }
 
-  update(part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
-    // If we've already set up this particular iterable, we don't need
-    // to do anything.
-    if (value === this._value) {
-      return;
-    }
-    this._value = value;
-    this.__iterate(part, mapper);
-    return noChange;
-  }
-
-  // Separate function to avoid an iffe in update; update can't be async
-  // because its return value must be `noChange`
-  private async __iterate(part: ChildPart, mapper?: Mapper<unknown>) {
-    let i = 0;
-    const {_value: value} = this;
-    for await (let v of value!) {
-      // Check to make sure that value is the still the current value of
-      // the part, and if not bail because a new value owns this part
-      if (this._value !== value) {
-        break;
-      }
-
-      // If we were disconnected, pause until reconnected
-      if (this._reconnectPromise) {
-        await this._reconnectPromise;
-      }
-
-      // When we get the first value, clear the part. This lets the
-      // previous value display until we can replace it.
-      if (i === 0) {
-        clearPart(part);
-      }
-      // As a convenience, because functional-programming-style
-      // transforms of iterables and async iterables requires a library,
-      // we accept a mapper function. This is especially convenient for
-      // rendering a template for each item.
-      if (mapper !== undefined) {
-        v = mapper(v, i);
-      }
-      const newPart = insertPart(part);
-      setChildPartValue(newPart, v);
-      i++;
+  // Override point for AsyncReplace to replace rather than append
```

Comment:
I made `AsyncAppend` subclass `AsyncReplace` since it's mostly identical code. As you can see in `AsyncAppend`, the only differences are:
* `validatePartType`: AsyncReplace can be used anywhere but AsyncAppend can only go in `ChildPart` position
* `update`: AsyncAppend needs to save the `__childPart` since it inserts new parts into it
* `commitValue`: AsyncAppend inserts new parts for each value, whereas AsyncReplace just calls `setValue` and overwrites the previous

Reply:
Oic, I had the comments backward. Fixed.